In [225]:
#Import libraries
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import os
import scipy.stats
%matplotlib inline

cwd = os.getcwd()
print(cwd)

#Change working directory
os.chdir(r'''C:\Users\wscott\AppData\Local\Programs\Python''')
print(cwd)

C:\Users\wscott\AppData\Local\Programs\Python
C:\Users\wscott\AppData\Local\Programs\Python


In [226]:
#Opens up the CSV file in my working directory and sets the dataframe as a callable variable
df = pd.read_csv('WELLCOME_APCspend2013_forThinkful.csv', encoding = "ISO-8859-1")

df

PMID/PMCID  \
0                                  NaN   
1                           PMC3679557   
2                23043264  PMC3506128    
3                  23438330 PMC3646402   
4                 23438216 PMC3601604    
5                           PMC3579457   
6                           PMC3709265   
7                 23057412 PMC3495574    
8                    PMCID: PMC3780468   
9                    PMCID: PMC3621575   
10                   PMCID: PMC3739413   
11                   PMCID: PMC3530961   
12                   PMCID: PMC3624797   
13                          PMC3413243   
14                          PMC3694353   
15                          PMC3572711   
16                            22610094   
17                   PMCID: PMC3586974   
18        23455506  PMCID: PMC3607399    
19          PMID: 24015914 PMC3833349    
20                       : PMC3805332    
21                                 NaN   
22            PMCID:\n    PMC3656742\n   
23                      PMCID: 3584654   
24                 22971149 PMC3466778   
25                            23072351   
26                          PMC3477822   
27                          PMC3530960   
28                23458425 PMC3630740    
29                      PMID: 20146481   
...                                ...   
2097                           3575693   
2098        PMID: 22188548 PMC3492912    
2099        PMID: 22764734 PMC3798095    
2100                    PMID: 23703895   
2101                 PMCID: PMC3664106   
2102                    PMID: 23649934   
2103                        PMC2955965   
2104        PMID: 19948006 PMC3551259    
2105                    PMID: 23305527   
2106                 PMCID: PMC3600532   
2107                        PMC3440570   
2108              23670821 PMC3738939    
2109                          23640856   
2110                    PMID: 23918316   
2111                 PMCID: PMC3728731   
2112       23495205685.88 PMC3743159     
2113         PMID:22997008 PMC3569869    
2114                           3708104   
2115                        PMC3713766   
2116                        PMC3819359   
2117                       PMC3773237    
2118                       PMC3815011    
2119                        PMC3805356   
2120                             Epub    
2121                        PMC3707567   
2122                           2901593   
2123                           3748854   
2124                           3785148   
2125          PMCID:\n    PMC3647051\n   
2126  PMID: 23846567  (Epub July 2013)   

                                         Publisher  \
0                                              CUP   
1                                              ACS   
2                                              ACS   
3                                              ACS   
4                                              ACS   
5                                              ACS   
6                                              ACS   
7                                              ACS   
8     ACS (Amercian Chemical Society) Publications   
9     ACS (Amercian Chemical Society) Publications   
10    ACS (Amercian Chemical Society) Publications   
11    ACS (Amercian Chemical Society) Publications   
12    ACS (Amercian Chemical Society) Publications   
13                                ACS Publications   
14                                ACS Publications   
15                                   AGA Institute   
16                                           AMBSB   
17           American Association of Immunologists   
18           American Association of Immunologists   
19                       American Chemical Society   
20                       American Chemical Society   
21                       American Chemical Society   
22                       American Chemical Society   
23                       AMERICAN CHEMICAL SOCIETY   
24                       American Chemical Society   
25                 

In [227]:
#Let's clean the 'Cost' column name
df = df.rename(columns={"COST (£) charged to Wellcome (inc VAT when charged)":"COST"})
df

PMID/PMCID  \
0                                  NaN   
1                           PMC3679557   
2                23043264  PMC3506128    
3                  23438330 PMC3646402   
4                 23438216 PMC3601604    
5                           PMC3579457   
6                           PMC3709265   
7                 23057412 PMC3495574    
8                    PMCID: PMC3780468   
9                    PMCID: PMC3621575   
10                   PMCID: PMC3739413   
11                   PMCID: PMC3530961   
12                   PMCID: PMC3624797   
13                          PMC3413243   
14                          PMC3694353   
15                          PMC3572711   
16                            22610094   
17                   PMCID: PMC3586974   
18        23455506  PMCID: PMC3607399    
19          PMID: 24015914 PMC3833349    
20                       : PMC3805332    
21                                 NaN   
22            PMCID:\n    PMC3656742\n   
23                      PMCID: 3584654   
24                 22971149 PMC3466778   
25                            23072351   
26                          PMC3477822   
27                          PMC3530960   
28                23458425 PMC3630740    
29                      PMID: 20146481   
...                                ...   
2097                           3575693   
2098        PMID: 22188548 PMC3492912    
2099        PMID: 22764734 PMC3798095    
2100                    PMID: 23703895   
2101                 PMCID: PMC3664106   
2102                    PMID: 23649934   
2103                        PMC2955965   
2104        PMID: 19948006 PMC3551259    
2105                    PMID: 23305527   
2106                 PMCID: PMC3600532   
2107                        PMC3440570   
2108              23670821 PMC3738939    
2109                          23640856   
2110                    PMID: 23918316   
2111                 PMCID: PMC3728731   
2112       23495205685.88 PMC3743159     
2113         PMID:22997008 PMC3569869    
2114                           3708104   
2115                        PMC3713766   
2116                        PMC3819359   
2117                       PMC3773237    
2118                       PMC3815011    
2119                        PMC3805356   
2120                             Epub    
2121                        PMC3707567   
2122                           2901593   
2123                           3748854   
2124                           3785148   
2125          PMCID:\n    PMC3647051\n   
2126  PMID: 23846567  (Epub July 2013)   

                                         Publisher  \
0                                              CUP   
1                                              ACS   
2                                              ACS   
3                                              ACS   
4                                              ACS   
5                                              ACS   
6                                              ACS   
7                                              ACS   
8     ACS (Amercian Chemical Society) Publications   
9     ACS (Amercian Chemical Society) Publications   
10    ACS (Amercian Chemical Society) Publications   
11    ACS (Amercian Chemical Society) Publications   
12    ACS (Amercian Chemical Society) Publications   
13                                ACS Publications   
14                                ACS Publications   
15                                   AGA Institute   
16                                           AMBSB   
17           American Association of Immunologists   
18           American Association of Immunologists   
19                       American Chemical Society   
20                       American Chemical Society   
21                       American Chemical Society   
22                       American Chemical Society   
23                       AMERICAN CHEMICAL SOCIETY   
24                       American Chemical Society   
25                 

In [228]:
#Count the each time the Journal title column was named
df.groupby('Journal title').count().sort_values(by=['Publisher'], ascending=False)

PMID/PMCID  Publisher  \
Journal title                                                               
PLoS One                                                    91         92   
PLoS ONE                                                    62         62   
Journal of Biological Chemistry                             47         48   
Nucleic Acids Research                                      20         21   
Proceedings of the National Academy of Sciences             19         19   
PLoS Neglected Tropical Diseases                            18         18   
Human Molecular Genetics                                    16         18   
Nature Communications                                       17         17   
PLoS Pathogens                                              15         15   
PLoS Genetics                                               14         15   
Neuroimage                                                  14         15   
Brain                                                       11         14   
BMC Public Health                                           12         14   
NeuroImage                                                  14         14   
PLOS ONE                                                    13         14   
Movement Disorders                                          11         13   
Developmental Cell                                           9         12   
Journal of Neuroscience                                      9         12   
Biochemical Journal                                         11         12   
Journal of General Virology                                 11         11   
Current Biology                                              9         10   
PLOS One                                                    10         10   
BMJ                                                          5         10   
Cell Reports                                                 9          9   
Neuron                                                       8          9   
Proceedings of the Royal Society B: Biological ...           8          8   
Journal of Cell Science                                      7          8   
Journal of Virology                                          7          8   
Neurobiology of Aging                                        8          8   
Journal of Structural Biology                                8          8   
...                                                        ...        ...   
Inflammatory Bowel Diseases                                  1          1   
Influenza Other Respi Viruses.                               1          1   
Int J Epidemiol                                              1          1   
Int Journal for Parasitology                                 1          1   
Heptology                                                    1          1   
Heart                                                        1          1   
Frontiers in Schizophrenia                                   1          1   
Health: An Interdisciplinary Journal for the So...           1          1   
Frontiers in Systems Neuroscience                            1          1   
Frontiers in T Cell Biology                                  1          1   
Fungal Biology                                               1          1   
Fungal Ecology                                               1          1   
Future Microbiology                                          1          1   
Future Neurology                                             1          1   
G3: Genes, Genomes, Genetics                                 1          1   
GENE                                                         1          1   
GENETICS                                                     1          1   
Genetic Epidemology                                          1          1   
Genetics                                                     1          1   
HBM JNL Human Brain Mapping                     

With all of the different spellings for PLos One, it's probably best to lowercase all of the strings to bucket the journals accordingly.

In [229]:
#Lower case the column and place it in a new column
lower = df['Journal title'].str.lower()
df['lc_journal_title'] = lower


In [230]:
df.groupby('lc_journal_title').count().sort_values(by=['Publisher'], ascending=False)

PMID/PMCID  Publisher  \
lc_journal_title                                                            
plos one                                                   188        190   
journal of biological chemistry                             52         53   
neuroimage                                                  28         29   
plos genetics                                               23         24   
plos pathogens                                              24         24   
nucleic acids research                                      22         23   
proceedings of the national academy of sciences             20         20   
plos neglected tropical diseases                            20         20   
nature communications                                       19         19   
human molecular genetics                                    17         19   
brain                                                       11         14   
bmc public health                                           12         14   
movement disorders                                          11         13   
developmental cell                                           9         12   
journal of neuroscience                                      9         12   
biochemical journal                                         11         12   
journal of general virology                                 11         11   
current biology                                             10         11   
bmj                                                          5         10   
neuron                                                       8          9   
cell reports                                                 9          9   
plosone                                                      9          9   
plos computational biology                                   8          9   
bmj open                                                     9          9   
european journal of immunology                               8          8   
proceedings of the royal society b: biological ...           8          8   
journal of clinical endocrinology and metabolism             8          8   
journal of cell science                                      7          8   
development                                                  7          8   
malaria journal                                              8          8   
...                                                        ...        ...   
immnunobiology                                               1          1   
immunological reviews                                        0          1   
immunology letters                                           1          1   
implement sci.                                               1          1   
infection genetics and evolution                             1          1   
inflammation research                                        1          1   
heart                                                        1          1   
health psychology                                            1          1   
frontiers in perception science                              1          1   
health promotion international                               0          1   
frontiers in schizophrenia                                   1          1   
frontiers in systems neuroscience                            1          1   
frontiers in t cell biology                                  1          1   
fungal biology                                               1          1   
fungal ecology                                               1          1   
future microbiology                                          1          1   
future neurology                                             1          1   
g3: genes, genomes, genetics                                 1          1   
genesis: journal of genetics                                 1          1   
genetic epidemology                             

Looks like the 5 most common journals are PLoS One, Journal of Biological, Neuroimage, PLos Genetics, and PLos Pathogens.

In [231]:
#Split it on '£' and turn the remaining string into a float to manipulate later
currency_split = df['COST'].str.split('£', expand=True)
df['Cost'] = currency_split[1].astype(float)
df

PMID/PMCID  \
0                                  NaN   
1                           PMC3679557   
2                23043264  PMC3506128    
3                  23438330 PMC3646402   
4                 23438216 PMC3601604    
5                           PMC3579457   
6                           PMC3709265   
7                 23057412 PMC3495574    
8                    PMCID: PMC3780468   
9                    PMCID: PMC3621575   
10                   PMCID: PMC3739413   
11                   PMCID: PMC3530961   
12                   PMCID: PMC3624797   
13                          PMC3413243   
14                          PMC3694353   
15                          PMC3572711   
16                            22610094   
17                   PMCID: PMC3586974   
18        23455506  PMCID: PMC3607399    
19          PMID: 24015914 PMC3833349    
20                       : PMC3805332    
21                                 NaN   
22            PMCID:\n    PMC3656742\n   
23                      PMCID: 3584654   
24                 22971149 PMC3466778   
25                            23072351   
26                          PMC3477822   
27                          PMC3530960   
28                23458425 PMC3630740    
29                      PMID: 20146481   
...                                ...   
2097                           3575693   
2098        PMID: 22188548 PMC3492912    
2099        PMID: 22764734 PMC3798095    
2100                    PMID: 23703895   
2101                 PMCID: PMC3664106   
2102                    PMID: 23649934   
2103                        PMC2955965   
2104        PMID: 19948006 PMC3551259    
2105                    PMID: 23305527   
2106                 PMCID: PMC3600532   
2107                        PMC3440570   
2108              23670821 PMC3738939    
2109                          23640856   
2110                    PMID: 23918316   
2111                 PMCID: PMC3728731   
2112       23495205685.88 PMC3743159     
2113         PMID:22997008 PMC3569869    
2114                           3708104   
2115                        PMC3713766   
2116                        PMC3819359   
2117                       PMC3773237    
2118                       PMC3815011    
2119                        PMC3805356   
2120                             Epub    
2121                        PMC3707567   
2122                           2901593   
2123                           3748854   
2124                           3785148   
2125          PMCID:\n    PMC3647051\n   
2126  PMID: 23846567  (Epub July 2013)   

                                         Publisher  \
0                                              CUP   
1                                              ACS   
2                                              ACS   
3                                              ACS   
4                                              ACS   
5                                              ACS   
6                                              ACS   
7                                              ACS   
8     ACS (Amercian Chemical Society) Publications   
9     ACS (Amercian Chemical Society) Publications   
10    ACS (Amercian Chemical Society) Publications   
11    ACS (Amercian Chemical Society) Publications   
12    ACS (Amercian Chemical Society) Publications   
13                                ACS Publications   
14                                ACS Publications   
15                                   AGA Institute   
16                                           AMBSB   
17           American Association of Immunologists   
18           American Association of Immunologists   
19                       American Chemical Society   
20                       American Chemical Society   
21                       American Chemical Society   
22                       American Chemical Society   
23                       AMERICAN CHEMICAL SOCIETY   
24                       American Chemical Society   
25                 

In [232]:
display(df.Cost.describe())

count      2114.000000
mean      24206.441892
std      147300.992149
min           0.000000
25%        1280.000000
50%        1889.950000
75%        2322.315000
max      999999.000000
Name: Cost, dtype: float64

In [233]:
display(df[['Cost', 'lc_journal_title']].groupby('lc_journal_title').describe())

Cost                 \
                                                   count           mean   
lc_journal_title                                                          
academy of nutrition and dietetics                   1.0    2379.540000   
acs chemical biology                                 5.0    1418.186000   
acs chemical neuroscience                            1.0    1186.800000   
acs nano                                             2.0     668.140000   
acta crystallographica section d,  biological c...   1.0     771.420000   
acta crystallographica section d: biological cr...   1.0     773.740000   
acta crystallographica section f: structural bi...   2.0     796.635000   
acta crystallographica, section d                    1.0     757.180000   
acta crystallography d                               1.0     774.190000   
acta d                                               1.0     750.160000   
acta dermato venereologica                           1.0     653.960000   
acta diabetologica                                   1.0    2336.280000   
acta f                                               1.0     754.900000   
acta neuropathol                                     1.0    1901.040000   
acta neuropathologica                                3.0    2161.063333   
acta opthalmologica                                  1.0    2270.160000   
acta physiol                                         1.0    1991.500000   
addiction                                            2.0    2136.225000   
advances in experimental medicine and biology        4.0    1928.457500   
age                                                  2.0    2170.140000   
age and ageing                                       1.0    2040.000000   
ageing & society                                     1.0    1695.000000   
aging cell                                           1.0    1970.570000   
aids                                                 3.0    2059.306667   
aids behav                                           1.0    1834.770000   
aids care                                            4.0    2294.225000   
aids journal                                         1.0    2015.720000   
aids research and therapy                            1.0    1240.000000   
aids uk                                              2.0    1836.920000   
alcohol and alcoholism                               1.0    2040.000000   
...                                                  ...            ...   
traffic                                              6.0    2129.338333   
transactions of the royal society of tropical m...   1.0    1524.080000   
translational psychiatry                             2.0    2880.000000   
trends in cognitive science                          1.0    2340.630000   
trends in genetics                                   2.0    2370.935000   
trends in microbiology                               1.0    2291.330000   
trends in molecular medicine                         1.0    2248.330000   
trends in neuroscience                               1.0    2402.810000   
trends in neurosciences                              2.0  501212.980000   
trends in parasitology                               4.0  251692.990000   
trials                                               4.0  250908.975000   
trials                                               1.0    1165.000000   
trop med int health                                  1.0    1834.770000   
tropical animal health & production                  1.0    2054.780000   
tropical medicine and international health           6.0    1843.163333   
tuberculosis                                         2.0    2161.255000   
urban history                                        1.0    2034.000000   
urolithiasis                                         1.0    1112.400000   
vaccine                                              5.0    1623.394000   
vascular pharmacology                                1.0    2352.530000   
ve

In [234]:
outlier_mask = df.Cost > 100000

In [235]:
df[outlier_mask]

PMID/PMCID  \
149                                         PMC3234811   
227                                            3708772   
277                                        PMC3668259    
358                                         PMC3219211   
404                                         PMC3533396   
410                                                NaN   
491                                  PMCID: PMC3464430   
560                                         PMC3632754   
630    Epub ahead of print April 2013 - print in press   
660                           PMID:23291342 PMC3581773   
669                                         PMC3594749   
670                                  PMCID: PMC3679449   
811                                                NaN   
815                                                NaN   
825                                                NaN   
829                              23200744  PMC3552157    
873                                     PMC3644713\n\n   
945                                     PMID: 22547635   
996                                        PMC3785137    
1036                    PMID: 23703681 (May 2013 Epub)   
1046                                        PMC3709478   
1065                                 PMCID: PMC3615658   
1138                                        PMC3510758   
1208  Epub ahead of print pub Jan 2013, print in press   
1256                               23981980 PMC3809720   
1309                                           2294692   
1367                            PMCID:    PMC3531293\n   
1386                                        PMC3527207   
1387                                               NaN   
1470                                           3547931   
1560                                           3726600   
1561                                          23285297   
1562                                       PMC3720940    
1563                                 PMCID: PMC3748079   
1564                            PMCID:\n    PMC3579793   
1565                            PMCID:\n    PMC3732248   
1566                          PMCID:\n    PMC3767740\n   
1597                                 PMCID: PMC3749955   
1598                          PMCID:\n    PMC3764210\n   
1612                                      PMC3777952\n   
1614                                       PMC 3712914   
1616                                       PMC 3460945   
1664                                (PMCID:PMC3757918)   
1675                                 Not yet available   
1835                                  PMCID:PMC3268392   
1939                                               NaN   
1947                                               NaN   
1961                                        PMC3557719   
1987                                    PMC3664409\n\n   

                                              Publisher  \
149                                               ASBMB   
227                                     BioMed Central    
277                                                 BMC   
358                          Cambridge University Press   
404                               Company of Biologists   
410                                                 CUP   
491                                            Elsevier   
560                                            Elsevier   
630                                            Elsevier   
660                                            Elsevier   
669                                            Elsevier   
670                                            Elsevier   
811                                            Elsevier   
815                                            Elsevier   
825                                            Elsevier   
829                                            Elsevier   
873                                           Frontiers   
945                                   John Wiley & Sons   
996                   

In [304]:
#Journal_Article = df[['Cost', 'lc_journal_title', 'Article title' ]].groupby('lc_journal_title')

In [305]:
#display(df[['Cost', 'lc_journal_title']].groupby('lc_journal_title').std())

In [306]:
#Journal_Article.std()

In [307]:
#STD_greater_than_mean = Journal_Article.std() > Journal_Article.mean()
#STD_greater_than_mean

In [308]:
#Journal_Article[STD_greater_than_mean]

So we want to look at all the columns that have the same 'Journal title.' After collecting the costs for each entry, if any of the entries are above the specific journal title's 95th percentile, change it to that 95th percentile (Winsorizing).

In [236]:
plos_one = (df.lc_journal_title == 'plos one')

In [310]:

plos_one_Cost = df[plos_one]['Cost']
plos_one_Cost

1282      1001.03
1283      1004.15
1284      1011.45
1285      1011.45
1286      1015.73
1287      1023.41
1288      1039.87
1289      1061.24
1290      1061.24
1291      1061.24
1292      1080.00
1293       794.93
1294       809.29
1295       819.34
1296       854.96
1297       901.50
1298       902.12
1299       903.89
1303      1061.93
1304       389.73
1305       443.38
1414      1002.86
1415      1005.86
1416      1005.96
1417      1008.97
1418      1009.98
1419      1010.80
1420      1013.03
1421      1019.76
1422      1027.73
          ...    
1553       896.32
1554       896.96
1555       896.99
1556       897.19
1557       897.61
1558       897.61
1559       899.34
1560    999999.00
1561    999999.00
1562    999999.00
1563    999999.00
1564    999999.00
1565    999999.00
1566    999999.00
1567       900.14
1568       904.34
1569       907.08
1570       908.01
1571       909.99
1572       910.20
1573       913.36
1574       915.53
1575       918.62
1576       953.48
1577      

In [238]:
np.percentile(plos_one_Cost, 95)

1761.7749999999996

In [239]:
plos_one_winsor_cost = scipy.stats.mstats.winsorize(plos_one_Cost, limits = 0.05)
plos_one_winsor_cost

masked_array(data=[1001.03, 1004.15, 1011.45, 1011.45, 1015.73, 1023.41,
                   1039.87, 1061.24, 1061.24, 1061.24, 1080.  ,  794.93,
                    809.29,  819.34,  854.96,  901.5 ,  902.12,  903.89,
                   1061.93,  773.45,  773.45, 1002.86, 1005.86, 1005.96,
                   1008.97, 1009.98, 1010.8 , 1013.03, 1019.76, 1027.73,
                   1034.95, 1036.38, 1036.78, 1040.21, 1040.7 , 1040.75,
                   1040.87, 1041.66, 1044.35, 1044.55, 1044.73, 1044.74,
                   1046.74, 1047.02, 1047.25, 1047.25, 1048.79, 1056.24,
                   1057.92, 1060.  , 1060.06, 1063.55, 1063.69, 1063.69,
                   1064.04, 1064.04, 1064.22, 1065.58, 1068.11, 1068.12,
                   1068.12, 1068.19, 1068.23, 1072.85, 1072.92, 1075.98,
                   1076.94, 1079.64, 1085.61, 1331.13, 1395.78, 1541.48,
                   1692.  , 1745.  , 1775.5 , 1775.5 ,  773.45, 1775.5 ,
                    773.45,  773.45,  773.45,  773.

In [240]:
df['test_col'] = df['Cost']
df

PMID/PMCID  \
0                                  NaN   
1                           PMC3679557   
2                23043264  PMC3506128    
3                  23438330 PMC3646402   
4                 23438216 PMC3601604    
5                           PMC3579457   
6                           PMC3709265   
7                 23057412 PMC3495574    
8                    PMCID: PMC3780468   
9                    PMCID: PMC3621575   
10                   PMCID: PMC3739413   
11                   PMCID: PMC3530961   
12                   PMCID: PMC3624797   
13                          PMC3413243   
14                          PMC3694353   
15                          PMC3572711   
16                            22610094   
17                   PMCID: PMC3586974   
18        23455506  PMCID: PMC3607399    
19          PMID: 24015914 PMC3833349    
20                       : PMC3805332    
21                                 NaN   
22            PMCID:\n    PMC3656742\n   
23                      PMCID: 3584654   
24                 22971149 PMC3466778   
25                            23072351   
26                          PMC3477822   
27                          PMC3530960   
28                23458425 PMC3630740    
29                      PMID: 20146481   
...                                ...   
2097                           3575693   
2098        PMID: 22188548 PMC3492912    
2099        PMID: 22764734 PMC3798095    
2100                    PMID: 23703895   
2101                 PMCID: PMC3664106   
2102                    PMID: 23649934   
2103                        PMC2955965   
2104        PMID: 19948006 PMC3551259    
2105                    PMID: 23305527   
2106                 PMCID: PMC3600532   
2107                        PMC3440570   
2108              23670821 PMC3738939    
2109                          23640856   
2110                    PMID: 23918316   
2111                 PMCID: PMC3728731   
2112       23495205685.88 PMC3743159     
2113         PMID:22997008 PMC3569869    
2114                           3708104   
2115                        PMC3713766   
2116                        PMC3819359   
2117                       PMC3773237    
2118                       PMC3815011    
2119                        PMC3805356   
2120                             Epub    
2121                        PMC3707567   
2122                           2901593   
2123                           3748854   
2124                           3785148   
2125          PMCID:\n    PMC3647051\n   
2126  PMID: 23846567  (Epub July 2013)   

                                         Publisher  \
0                                              CUP   
1                                              ACS   
2                                              ACS   
3                                              ACS   
4                                              ACS   
5                                              ACS   
6                                              ACS   
7                                              ACS   
8     ACS (Amercian Chemical Society) Publications   
9     ACS (Amercian Chemical Society) Publications   
10    ACS (Amercian Chemical Society) Publications   
11    ACS (Amercian Chemical Society) Publications   
12    ACS (Amercian Chemical Society) Publications   
13                                ACS Publications   
14                                ACS Publications   
15                                   AGA Institute   
16                                           AMBSB   
17           American Association of Immunologists   
18           American Association of Immunologists   
19                       American Chemical Society   
20                       American Chemical Society   
21                       American Chemical Society   
22                       American Chemical Society   
23                       AMERICAN CHEMICAL SOCIETY   
24                       American Chemical Society   
25                 

In [241]:
df.loc[plos_one, 'test_col'] = plos_one_winsor_cost

In [242]:
df[plos_one]

PMID/PMCID                         Publisher  \
1282                      3517619                              PLoS   
1283                      3498109                              PLoS   
1284                      3515553                              PLoS   
1285                      3522679                              PLoS   
1286                      3485223                              PLoS   
1287                   PMC3547059                              PLoS   
1288                      3573029                              PLoS   
1289                      3769269                              PLoS   
1290                      3782430                              PLoS   
1291                      3797097                              PLoS   
1292                      3675909                              PLoS   
1293                   PMC3566186                              PLOS   
1294                   PMC3694955                              PLoS   
1295                   PMC3729655                              PLOS   
1296                   PMC3774797                              PLoS   
1297                      3620162                              PLoS   
1298                   PMC3694164                              PLOS   
1299                      3669132                              Plos   
1303             PMCID:PMC3711905  PLoS (Public Library of Science)   
1304            PMCID: PMC3661540  PLoS (Public Library of Science)   
1305             PMCID:PMC3706442  PLoS (Public Library of Science)   
1414                   PMC3443216         Public Library of Science   
1415                   PMC3557262         Public Library of Science   
1416            PMCID: PMC3559699         Public Library of Science   
1417            PMCID: PMC3502435         Public Library of Science   
1418           PMCID: PMC3491040          Public Library of Science   
1419       PMCID:\n    PMC3532161         Public Library of Science   
1420                   PMC3474824         Public Library of Science   
1421                   PMC3509064         Public Library of Science   
1422                     23285115         Public Library of Science   
...                           ...                               ...   
1553           PMCID: PMC3634070          Public Library of Science   
1554     PMCID:\n    PMC3496742\n         Public Library of Science   
1555                   PMC3591389         Public Library of Science   
1556            PMCID: PMC3606301         Public Library of Science   
1557                   PMC3637166         Public Library of Science   
1558                   PMC3655936         Public Library of Science   
1559                   PMC3590148         Public Library of Science   
1560                      3726600         Public Library of Science   
1561                     23285297         Public Library of Science   
1562                  PMC3720940          Public Library of Science   
1563            PMCID: PMC3748079         Public Library of Science   
1564       PMCID:\n    PMC3579793         Public Library of Science   
1565       PMCID:\n    PMC3732248         Public Library of Science   
1566     PMCID:\n    PMC3767740\n         Public Library of Science   
1567                   PMC3642175         Public Library of Science   
1568                   PMC3594159         Public Library of Science   
1569                   PMC3691179         Public Library of Science   
1570            PMCID: PMC3606242         Public Library of Science   
1571            PMCID: PMC3694102         Public Library of Science   
1572                   PMC3475710         Public Library of Science   
1573       PMCID:\n    PMC3636223         Public Library of Science   
1574            PMCID: PMC3720700         Public Library of Science   
1575         23805259 PMC3689675          Public Library of Science   
1576           PMCID: PMC3790696          Public Library of Science   
1577     PMCID:\n    PMC3609862\n         Publi

In [203]:
#lc_journal_title_list = df['lc_journal_title'].unique()
#lc_journal_title_list[0]

'psychological medicine'

Test Algorithm:

plos_one = (df.lc_journal_title == 'plos one')
plos_one_Cost = df[plos_one]['Cost']
plos_one_Cost
plos_one_winsor_cost = scipy.stats.mstats.winsorize(plos_one_Cost, limits = 0.05)
plos_one_winsor_cost
df.loc[plos_one, 'test_col'] = plos_one_winsor_cost

In [269]:
df['winsor_cost'] = df.Cost

#for current_journal in range(0,20):     
for current_journal in range(0,lc_journal_title_list.size):
    if type(lc_journal_title_list[current_journal]) is str:
    #isinstance(lc_journal_title_list[current_journal], basestring):
        journal_mask = (df.lc_journal_title == lc_journal_title_list[current_journal])
        journal_Cost = df[journal_mask]['Cost']
        journal_winsor_cost = scipy.stats.mstats.winsorize(journal_Cost, limits = 0.05)
        df.loc[journal_mask, 'winsor_cost'] = journal_winsor_cost
        test_set = set(journal_winsor_cost).symmetric_difference(set(journal_Cost))
        if len(test_set) != 0:
            print(str(lc_journal_title_list[current_journal]) + ' has been changed.')
    else: 
        print(str(current_journal) + ': ' + str(lc_journal_title_list[current_journal]) + ' (Journal Title Blank)') 

journal of biological chemistry has been changed.
bmc biology has been changed.
bmc genetics has been changed.
bmc genome biology has been changed.
bmc genomics has been changed.
biophysical journal has been changed.
proceedings of the national academy of sciences has been changed.
neuroimage has been changed.
506: nan (Journal Title Blank)
nucleic acids research has been changed.
plos genetics has been changed.
plos one has been changed.
plos pathogens has been changed.
plos neglected tropical diseases has been changed.


In [303]:
testing = (df.Cost > 100000 )
df[testing]

PMID/PMCID  \
149                                         PMC3234811   
227                                            3708772   
277                                        PMC3668259    
358                                         PMC3219211   
404                                         PMC3533396   
410                                                NaN   
491                                  PMCID: PMC3464430   
560                                         PMC3632754   
630    Epub ahead of print April 2013 - print in press   
660                           PMID:23291342 PMC3581773   
669                                         PMC3594749   
670                                  PMCID: PMC3679449   
811                                                NaN   
815                                                NaN   
825                                                NaN   
829                              23200744  PMC3552157    
873                                     PMC3644713\n\n   
945                                     PMID: 22547635   
996                                        PMC3785137    
1036                    PMID: 23703681 (May 2013 Epub)   
1046                                        PMC3709478   
1065                                 PMCID: PMC3615658   
1138                                        PMC3510758   
1208  Epub ahead of print pub Jan 2013, print in press   
1256                               23981980 PMC3809720   
1309                                           2294692   
1367                            PMCID:    PMC3531293\n   
1386                                        PMC3527207   
1387                                               NaN   
1470                                           3547931   
1560                                           3726600   
1561                                          23285297   
1562                                       PMC3720940    
1563                                 PMCID: PMC3748079   
1564                            PMCID:\n    PMC3579793   
1565                            PMCID:\n    PMC3732248   
1566                          PMCID:\n    PMC3767740\n   
1597                                 PMCID: PMC3749955   
1598                          PMCID:\n    PMC3764210\n   
1612                                      PMC3777952\n   
1614                                       PMC 3712914   
1616                                       PMC 3460945   
1664                                (PMCID:PMC3757918)   
1675                                 Not yet available   
1835                                  PMCID:PMC3268392   
1939                                               NaN   
1947                                               NaN   
1961                                        PMC3557719   
1987                                    PMC3664409\n\n   

                                              Publisher  \
149                                               ASBMB   
227                                     BioMed Central    
277                                                 BMC   
358                          Cambridge University Press   
404                               Company of Biologists   
410                                                 CUP   
491                                            Elsevier   
560                                            Elsevier   
630                                            Elsevier   
660                                            Elsevier   
669                                            Elsevier   
670                                            Elsevier   
811                                            Elsevier   
815                                            Elsevier   
825                                            Elsevier   
829                                            Elsevier   
873                                           Frontiers   
945                                   John Wiley & Sons   
996                   

In [314]:
test_winsor = [1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,2,50000000]

In [315]:
test_winsor

[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 50000000]

In [316]:
scipy.stats.mstats.winsorize(test_winsor, limits = 0.05)

masked_array(data=[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
                   2, 2],
             mask=False,
       fill_value=999999)

In [245]:
display(df[['winsor_cost', 'lc_journal_title']].groupby('lc_journal_title').describe())

winsor_cost                 \
                                                         count           mean   
lc_journal_title                                                                
academy of nutrition and dietetics                         1.0    2379.540000   
acs chemical biology                                       5.0    1418.186000   
acs chemical neuroscience                                  1.0    1186.800000   
acs nano                                                   2.0     668.140000   
acta crystallographica section d,  biological c...         1.0     771.420000   
acta crystallographica section d: biological cr...         1.0     773.740000   
acta crystallographica section f: structural bi...         2.0     796.635000   
acta crystallographica, section d                          1.0     757.180000   
acta crystallography d                                     1.0     774.190000   
acta d                                                     1.0     750.160000   
acta dermato venereologica                                 1.0     653.960000   
acta diabetologica                                         1.0    2336.280000   
acta f                                                     1.0     754.900000   
acta neuropathol                                           1.0    1901.040000   
acta neuropathologica                                      3.0    2161.063333   
acta opthalmologica                                        1.0    2270.160000   
acta physiol                                               1.0    1991.500000   
addiction                                                  2.0    2136.225000   
advances in experimental medicine and biology              4.0    1928.457500   
age                                                        2.0    2170.140000   
age and ageing                                             1.0    2040.000000   
ageing & society                                           1.0    1695.000000   
aging cell                                                 1.0    1970.570000   
aids                                                       3.0    2059.306667   
aids behav                                                 1.0    1834.770000   
aids care                                                  4.0    2294.225000   
aids journal                                               1.0    2015.720000   
aids research and therapy                                  1.0    1240.000000   
aids uk                                                    2.0    1836.920000   
alcohol and alcoholism                                     1.0    2040.000000   
...                                                        ...            ...   
traffic                                                    6.0    2129.338333   
transactions of the royal society of tropical m...         1.0    1524.080000   
translational psychiatry                                   2.0    2880.000000   
trends in cognitive science                                1.0    2340.630000   
trends in genetics                                         2.0    2370.935000   
trends in microbiology                                     1.0    2291.330000   
trends in molecular medicine                               1.0    2248.330000   
trends in neuroscience                                     1.0    2402.810000   
trends in neurosciences                                    2.0  501212.980000   
trends in parasitology                                     4.0  251692.990000   
trials                                                     4.0  250908.975000   
trials                                                     1.0    1165.000000   
trop med int health                                        1.0    1834.770000   
tropical animal health & production                        1.0    2054.780000   
tropical medicine and international health                 6.0    1843.163333   
tuberculosis                                               2.0    2161.255000   
u

In [320]:
bio_chem = df[(df.lc_journal_title == 'journal of biological chemistry')]
bio_chem

PMID/PMCID  \
16                     22610094   
60                   PMC3576085   
61         23239883 PMC3561570    
64            PMCID: PMC3642348   
65   PMID: 22992744  PMC3493908   
66            PMCID: PMC3531748   
67          PMCID: PMC3436574\n   
68   PMID: 23223336 PMC3543027    
69            PMCID: PMC3795252   
70            PMCID: PMC3724652   
71     PMCID:\n    PMC3597794\n   
72                   PMC3522294   
73                   PMC3774408   
74                   PMC3610987   
75                      3610976   
76               PMC3696691\n\n   
77        23344959  PMC3585060    
78            PMCID: PMC3481303   
79       PMCID:\n    PMC3789947   
121        23129765 PMC3531760    
122                  PMC3516716   
123                 PMC3581394    
124                 PMC3554938    
125                  PMC2757965   
126                         NaN   
127                     3234717   
128                  PMC3173209   
129                     3696646   
130              PMCID: 3682544   
131                     3431625   
132                  PMC3488092   
133         2339840 PMC3591641    
134                  PMC3630886   
135                     2952269   
136                     3121340   
137                  PMC3567677   
138                 PMC3774368    
139                  PMC3630861   
140         23569211 PMC3668706   
141                  PMC3488048   
142                  PMC3724654   
143                     2797226   
144                  PMC3442528   
145                     2898433   
146                     3397837   
147          PMCID: PMC3750171    
148                     2975197   
149                  PMC3234811   
150                     3137072   
151                  PMC3522295   
152               : PMC3711325    
163                 PMC3757190    
341        23264619 PMC3576117    

                                             Publisher  \
16                                               AMBSB   
60   American Soc for Biochemistry and Molecular Bi...   
61   American Society for Biochemistry and Molecula...   
64   American Society for Biochemistry and Molecula...   
65   American Society for Biochemistry and Molecula...   
66   American Society for Biochemistry and Molecula...   
67   American Society for Biochemistry and Molecula...   
68   American Society for Biochemistry and Molecula...   
69   American Society for Biochemistry and Molecula...   
70   American Society for Biochemistry and Molecula...   
71   American Society for Biochemistry and Molecula...   
72   American Society for Biochemistry and Molecula...   
73   American Society for Biochemistry and Molecula...   
74   American Society for Biochemistry and Molecula...   
75   American Society for Biochemistry and Molecula...   
76   American Society for Biochemistry and Molecula...   
77   American Society for Biochemistry and Molecula...   
78   American Society for Biochemistry and Molecula...   
79   American Society for Biochemistry and Molecula...   
121                                              ASBMB   
122                                              ASBMB   
123                                              ASBMB   
124                                              ASBMB   
125                                              ASBMB   
126                                              ASBMB   
127                                              ASBMB   
128                                              ASBMB   
129                                              ASBMB   
130                                              ASBMB   
131                                              ASBMB   
132                                              ASBMB   
133                                              ASBMB   
134                                              ASBMB   
135                                              ASBMB   
136                                              ASBMB   
137                                       

In [318]:
set_test = set(bio_chem['winsor_cost']).symmetric_difference(set(bio_chem['Cost']))
len(weak) == 0

False

In [319]:
set_test2 = set(bio_chem['winsor_cost']).symmetric_difference(set(bio_chem['winsor_cost']))
len(set_test2) == 0

True

In [325]:
display(df[['winsor_cost', 'lc_journal_title']].groupby('lc_journal_title').agg(['mean', 'median', 'count', 'std']))

winsor_cost              \
                                                             mean      median   
lc_journal_title                                                                
academy of nutrition and dietetics                    2379.540000    2379.540   
acs chemical biology                                  1418.186000    1294.590   
acs chemical neuroscience                             1186.800000    1186.800   
acs nano                                               668.140000     668.140   
acta crystallographica section d,  biological c...     771.420000     771.420   
acta crystallographica section d: biological cr...     773.740000     773.740   
acta crystallographica section f: structural bi...     796.635000     796.635   
acta crystallographica, section d                      757.180000     757.180   
acta crystallography d                                 774.190000     774.190   
acta d                                                 750.160000     750.160   
acta dermato venereologica                             653.960000     653.960   
acta diabetologica                                    2336.280000    2336.280   
acta f                                                 754.900000     754.900   
acta neuropathol                                      1901.040000    1901.040   
acta neuropathologica                                 2161.063333    2250.970   
acta opthalmologica                                   2270.160000    2270.160   
acta physiol                                          1991.500000    1991.500   
addiction                                             2136.225000    2136.225   
advances in experimental medicine and biology         1928.457500    1928.460   
age                                                   2170.140000    2170.140   
age and ageing                                        2040.000000    2040.000   
ageing & society                                      1695.000000    1695.000   
aging cell                                            1970.570000    1970.570   
aids                                                  2059.306667    1968.630   
aids behav                                            1834.770000    1834.770   
aids care                                             2294.225000    2316.010   
aids journal                                          2015.720000    2015.720   
aids research and therapy                             1240.000000    1240.000   
aids uk                                               1836.920000    1836.920   
alcohol and alcoholism                                2040.000000    2040.000   
...                                                           ...         ...   
traffic                                               2129.338333    2273.455   
transactions of the royal society of tropical m...    1524.080000    1524.080   
translational psychiatry                              2880.000000    2880.000   
trends in cognitive science                           2340.630000    2340.630   
trends in genetics                                    2370.935000    2370.935   
trends in microbiology                                2291.330000    2291.330   
trends in molecular medicine                          2248.330000    2248.330   
trends in neuroscience                                2402.810000    2402.810   
trends in neurosciences                             501212.980000  501212.980   
trends in parasitology                              251692.990000    2384.110   
trials                                              250908.975000    1253.250   
trials                                                1165.000000    1165.000   
trop med int health                                   1834.770000    1834.770   
tropical animal health & production                   2054.780000    2054.780   
tropical medicine and international health            1843.163333    1883.625   
tuberculosis                                          2161.255000    2161.255   
urba

Note, there are high standard deviations for journal titles that have less than 20 article entries. One option could be to automatically change all 999,999 entries to their max value, IF there are less than 20 articles. Otherwise, adhere to the Winsorizing method.